lib/amazonka/src/Amazonka/Env/Hooks.hs

```diff
@@ -0,0 +1,420 @@
+--
+-- -- Example: Run @aSignedRequestHook@ on anything that is a @Signed GetObjectRequest@:
+-- @
+-- requestHook (hookMatchingType_ @(Signed GetObjectRequest) aSignedRequestHook) :: Hooks -> Hooks
+-- @
+hookMatchingType_ ::
+  forall a b. (Typeable a, Typeable b) => Hook_ a -> Hook_ b -> Hook_ b
+hookMatchingType_ newHook oldHook = \env a -> do
+  oldHook env a
+  case eqT @a @b of
+    Just Refl -> newHook env a
+    Nothing -> pure ()
+
+-- | When @a@ and @b@ are the same type, do not call any more hooks.
+--
+-- -- Example: Prevent any request hooks from running against a @PutObjectRequest@:
+-- @
+-- requestHook (blockMatchingType @PutObjectRequest) :: Hooks -> Hooks
+-- @
+blockMatchingType :: forall a b. (Typeable a, Typeable b) => Hook b -> Hook b
+blockMatchingType oldHook = \env -> case eqT @a @b of
+  Just Refl -> pure
+  Nothing -> oldHook env
+
+-- | When @a@ and @b@ are the same type, do not call any more hooks.
+--
+-- -- Example: Prevent any error hooks from running against errors caused by a @PutObjectRequest@:
+-- @
+-- errorHook (blockMatchingType @(Request PutObjectRequest, Error)) :: Hooks -> Hooks
+-- @
+blockMatchingType_ :: forall a b. (Typeable a, Typeable b) => Hook_ b -> Hook_ b
+blockMatchingType_ oldHook = \env a -> case eqT @a @b of
+  Just Refl -> pure ()
+  Nothing -> oldHook env a
+
+-- | Run the wrapped hook unless the given 'Fold' or 'Traversal'
+-- matches the error. With the error matches in each service binding,
+-- you can selectively silence certain errors:
+-- @
+-- -- Assuming `env :: Amazonka.Env` and `putRequest :: DynamoDB.PutRequest`,
+-- -- this silences a single type of error for a single call:
+-- send (env & #hooks %~ errorHook (silenceError DynamoDB._ConditionalCheckFailedException))
+-- @
+silenceError ::
+  Getting (First e) Error e ->
+  Hook_ (Request a, Error) ->
+  Hook_ (Request a, Error)
+silenceError g oldHook = \env t@(_, err) ->
+  case err ^? g of
+    Nothing -> oldHook env t
+    Just _ -> pure ()
+
+-- | Add default logging hooks. The default 'Env'' from
+-- 'Amazonka.Env.newEnv' already has logging hooks installed, so you
+-- probably only want this if you are building your own 'Hooks' from
+-- scratch.
+addLoggingHooks :: Hooks -> Hooks
+addLoggingHooks
+  hooks@Hooks
+    { signedRequest,
+      clientRequest,
+      clientResponse,
+      requestRetry,
+      awaitRetry,
+      error
+    } =
+    hooks
+      { signedRequest = \env@Env {logger} s@Signed {signedMeta} -> do
+          signedRequest env s
+          logTrace logger signedMeta,
+        clientRequest = \env@Env {logger} rq -> do
+          rq' <- clientRequest env rq
+          rq' <$ logDebug logger rq',
+        clientResponse = \env@Env {logger} t@(_, rs) -> do
+          clientResponse env t
+          logDebug logger rs,
+        requestRetry = \env@Env {logger} t@(_, name, retryStatus) -> do
+          requestRetry env t
+          logDebug logger
+            . mconcat
+            . intersperse " "
```

Comment:
Since it's used twice here, is it worth defining a `unwords :: [Builder] -> Builder`?

Reply:
Yeah I'll do it, since it's likely to pay off if we add any more logging hooks (e.g., for `finalError`.)